buildSrc/src/main/java/com/google/firebase/gradle/plugins/FirebaseLibraryPlugin.kt

```diff
@@ -0,0 +1,150 @@
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.attributes.Attribute
+import org.gradle.api.publish.PublishingExtension
+import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.publish.tasks.GenerateModuleMetadata
+import org.gradle.kotlin.dsl.getByType
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+class FirebaseLibraryPlugin : Plugin<Project> {
+
+  override fun apply(project: Project) {
+    project.apply(ImmutableMap.of("plugin", "com.android.library"))
+    project.apply(ImmutableMap.of("plugin", LicenseResolverPlugin::class.java))
+    project.apply(ImmutableMap.of("plugin", "com.github.sherter.google-java-format"))
+    project.extensions.getByType(GoogleJavaFormatExtension::class.java).toolVersion = "1.10.0"
+    val firebaseLibrary =
+      project.extensions.create(
+        "firebaseLibrary",
+        FirebaseLibraryExtension::class.java,
+        project,
+        LibraryType.ANDROID
+      )
+    val android = project.extensions.getByType<LibraryExtension>()
+    android.compileOptions {
+      sourceCompatibility = JavaVersion.VERSION_1_8
+      targetCompatibility = JavaVersion.VERSION_1_8
+    }
+
+    // In the case of and android library signing config only affects instrumentation test APK.
+    // We need it signed with default debug credentials in order for FTL to accept the APK.
+    android.buildTypes { getByName("release").signingConfig = getByName("debug").signingConfig }
+
+    android.defaultConfig {
+      buildConfigField("String", "VERSION_NAME", "\"" + project.version + "\"")
+    }
+
+    // see https://github.com/robolectric/robolectric/issues/5456
+    android.testOptions {
+      unitTests.all {
+        it.systemProperty("robolectric.dependency.repo.id", "central")
+        it.systemProperty("robolectric.dependency.repo.url", "https://repo1.maven.org/maven2")
+        it.systemProperty("javax.net.ssl.trustStoreType", "JKS")
+      }
+    }
+
+    setupApiInformationAnalysis(project, android)
+    android.testServer(FirebaseTestServer(project, firebaseLibrary.testLab, android))
+    setupStaticAnalysis(project, firebaseLibrary)
+    configurePublishing(project, firebaseLibrary, android)
+
+    // reduce the likelihood of kotlin module files colliding.
+    project.tasks.withType(KotlinCompile::class.java) {
+      kotlinOptions.freeCompilerArgs = ImmutableList.of("-module-name", kotlinModuleName(project))
+    }
+
+    project.pluginManager.apply(DackkaPlugin::class.java)
+    project.pluginManager.apply(GitSubmodulePlugin::class.java)
+    project.tasks.getByName("preBuild").dependsOn("updateGitSubmodules")
+  }
+
+  private fun setupApiInformationAnalysis(project: Project, android: LibraryExtension) {
+    val srcDirs = android.sourceSets.getByName("main").java.srcDirs
+
+    val mainSourceSets = android.sourceSets.getByName("main")
+    val getKotlinDirectories = mainSourceSets::class.java.getDeclaredMethod("getKotlinDirectories")
```

Comment:
`srcDirs` should be the same. The convention is confusing, I had a discussion with one of the folks over at AndroidX regarding this exact scenario awhile ago.
```kotlin
    val srcDirs = android.sourceSets.getByName("main").java.srcDirs

    val apiInfo = getApiInfo(project, srcDirs)
    val generateApiTxt = getGenerateApiTxt(project, srcDirs)
    val docStubs = getDocStubs(project, srcDirs)
```

Reply:
This is how it originally was, however, it was causing issues, and Yifan's change is what made this pass the tests properly, which we need to talk to Vinay about.

Reply:
I am worried that this would only include kotlinSrcDirs and not all srcDirs. Is the job failing in master as well?

Reply:
Please check if ./gradlew firebase-storage:generateApiTextFile works fine
and ./gradlew firebase-storage:apiInfo.. And if it generates the same apiTextFile. I think the original code was written mainly for java src Dirs. Could you also check if the merge of android.sourceSets.getByName("main").java.srcDirs and  mainSourceSets::class.java.getDeclaredMethod("getKotlinDirectories") also fixes this job failure?

Reply:
Both tasks are running fine as far as I can tell with the output api.txt. After rerunning the tests it definitely doesn't seem like a flake, Yifan's merge fixed the job failure.

Reply:
Cool lgtm from myside then

Reply:
Did you try the merge approach and see if it fixes the issue as well?  union of android.sourceSets.getByName("main").java.srcDirs and mainSourceSets::class.java.getDeclaredMethod("getKotlinDirectories")